File: backend/app/schemas/table.py
"""Pydantic schemas for catalog tables.

Table existence comes from the external ``iceberg_table_metrics`` table; a
single table's columns are read live from Spark Connect on demand.
"""

from pydantic import BaseModel


class TableColumn(BaseModel):
    name: str
    data_type: str | None = None
    ordinal_position: int = 0


class TableRef(BaseModel):
    """A reference to a catalog table (no schema)."""

    namespace: str
    table_name: str


class TableListItem(BaseModel):
    namespace: str
    table_name: str


class TableListResponse(BaseModel):
    items: list[TableListItem]
    total: int


class TableSchema(BaseModel):
    """A table's live schema (columns) + its consume snippet."""

    namespace: str
    table_name: str
    columns: list[TableColumn] = []
    consume_snippet: str = ""


class NamespaceListResponse(BaseModel):
    items: list[str]

File: frontend/src/types/table.ts
import type { PipelineField } from "./pipeline";

export interface TableColumn {
  name: string;
  data_type: string | null;
  ordinal_position: number;
}

/** A reference to a catalog table (no schema). */
export interface TableRef {
  namespace: string;
  table_name: string;
}

export interface TableListItem {
  namespace: string;
  table_name: string;
}

export interface TableListResponse {
  items: TableListItem[];
  total: number;
}

export interface NamespaceListResponse {
  items: string[];
}

/** A table's live schema (columns) + consume snippet, fetched on demand. */
export interface TableSchema {
  namespace: string;
  table_name: string;
  columns: TableColumn[];
  consume_snippet: string;
}

/** Adapt catalog columns to the PipelineField shape SchemaViewer expects. */
export function columnsToFields(table: TableSchema): PipelineField[] {
  return table.columns.map((c) => ({
    id: `${table.namespace}.${table.table_name}.${c.name}`,
    name: c.name,
    data_type: c.data_type,
    ordinal_position: c.ordinal_position,
  }));
}
